LSPS4/README.md

```diff
@@ -0,0 +1,1923 @@
+    that trust).
+
+> **Rationale** As mentioned in the section "Trust Models" above,
+> in LSP trusts client model, the client is allowed to choose its
+> risk appetite; it may accept 0 confirmations, 1, 3, 6, 100, or
+> however many it believes would be safe with the particular LSP.
+
+As soon as both client and LSP have sent `channel_ready` for the
+new channel, the LSP then forwards PPPs in the PPP Opening Set to
+the client.
+
+The LSP MUST NOT forward other HTLCs to the new channel until
+the PPP Opening Set is empty.
+
+The LSP generates non-standard forwards, where the amount received
+by the client is smaller than specified in the onion;
+the client MUST accept the non-standard forward(s), provided they
+sum to at least `payment_size_msat - opening_fee`.
+
+The LSP MUST ensure that each forwarded part forwards at least
+`client_htlc_minimum_msat` millisatoshis.
+The LSP MUST ensure that all parts forwarded sum up to at least
+`payment_size_msat - opening_fee`.
+
+The LSP MUST include an `extra_fee` (type 65536) TLV to the
+`update_add_htlc` message of forwarded parts that have had
+fees deducted.
+The LSP MUST NOT include this `extra_fee` TLV if the part does
+not have fees deducted (for example, if this is a multipart
+payment and the other parts have already paid for the fee).
+The `extra_fee` TLV MUST have a length of 8, and the value is a
+big-endian 64-bit number containing the number of millisatoshis
+deducted from this part.
+
+The client MUST check that the sum of all `extra_fee`s is less
+than or equal to the `opening_fee`, and that the actual amount of
+each HTLC is equal to the expected amount in the onion minus the
+individual `extra_fee`.
+
+`extra_fee` is currently proposed in [bLIP-0025][].
+
+[bLIP-0025]: https://github.com/lightning/blips/pull/25
+
+Once the corresponding HTLC has been sent by the LSP and claimed
+by the client, the LSP removes the corresponding PPP from the
+PPP Opening Set.
+
+Once the PPP Opening Set is empty, the LSP removes the set,
+and the `lsps4.open_with_ppps` API can be invoked again.
+
+### 8.  Post-opening Normal Operation
+
+The client MUST NOT use the `alias` SCID sent in the 0-conf
+channel open (unless it is equal to the LSPS4 SCID, which
+is **not** assured), or the "real" SCID of the channel once
+confirmed.
+The client MUST use only the SCID issued via `lsps4.reserve_scid`
+for all invoices and offers it creates.
+
+The LSP SHOULD fail an incoming HTLC if it uses the `alias` or
+"real" SCID of the channel as the next hop, with the failure
+`unknown_next_peer`, if the `alias` is not the same as the
+LSPS4 SCID.
+
+> **Rationale** Some node software implementations do not have
+> an API to set the `alias` of the channel.
+> The LSPS4 SCID, in any case, represents all the channels
+> with the client, not any specific channel.
+> It also represents a possible future channel if there is no
+> actual channel between the LSP and the client yet.
```

Comment:
It could be useful for the client to have the option to bypass the LSPS4 spec by issuing invoices with the real alias scid. In that case it is just a normal forward, without the PPP overhead.

Reply:
Client can be simpler if it always just uses the LSPS4 SCID.  It can issue invoices with 0 regard for current inbound liquidity, or if it issues multiple invoices (some of which may or may not be paid, e.g. merchant application) and does not want to have to allocate inbound liquidity to invoices.

For example, suppose the client currently has 1 unit of inbound.  Then it issues two invoices of 1 unit each, with no assurance of whether or not it is going to get paid.  If it used the "real" SCID then if both invoices are paid, then the second one will fail due to lack of capacity.

An LSP can ameliorate this by putting the second one in a PPP even though it did not use the LSPS4 SCID.  This is equivalent to saying that "the real and/or alias SCID of a channel with a client is an alias of the LSPS4 SCID".  But it is much much simpler to just say "client always uses the LSPS4 SCID" for implementation.

Reply:
Say the client selects a new LSP to get its future channels from, it doesn't want new channels from the old LSP. It can still use the channel with the old LSP, by including a 'normal' hop hint in the invoice, and perhaps another hop hint for the new LSP.

Reply:
>Say the client selects a new LSP to get its future channels from, it doesn't want new channels from the old LSP. It can still use the channel with the old LSP, by including a 'normal' hop hint in the invoice, and perhaps another hop hint for the new LSP.

And the "normal" routehint in this case would be the result from `lsps4.get_scid`.